### PR TITLE
feat: add support for using custom .m2 directory

### PR DESCRIPTION
**Pull request by agdimech**
The existing Java packaging process uses a localRepository which points to a newly created tmp directory. This means that when building, all dependencies need to be freshly downloaded which is contributing a significant portion of time to the packaging process.

This PR provides the option to the user (hidden) to use a non-transient .m2 local repository if they so choose. I have experienced a huge improvement to performance by enabling this setting (5s with enabled vs 4 minutes using defaults).

---

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0

### CODE REVIEW

File: packages/jsii-pacmak/bin/jsii-pacmak.ts [hunk 1]
```diff
@@ -122,6 +122,12 @@ import { VERSION_DESC } from '../lib/version';
       // This is expected to be a path, which should be normalized
       normalize: true,
     })
+    .option('java-use-default-m2', {
+      type: 'boolean',
+      desc: 'Use default M2 repository when packaging a Java package.',
+      default: false,
```
Comment (agdimech) [inline packages/jsii-pacmak/bin/jsii-pacmak.ts:128]: I have set the default to `false` for this setting to maintain backward compatibility. I am wandering whether this should be defaulted to `true` in future as I am not sure what negative side effects there would be by using the user's .m2 repository.

**Comment by RomainMuller** [comment]
We moved away from the default cache locations for both .NET and Java because this had surprising consequences due to how both will always prefer artifacts already in cache (which is good), AND store the outputs of local builds into this cache (which is where surprises come from).

I reckon the performance improvement you see comes from not re-downloading dependencies on each build... which is also achieved when using a non-default, but also non-transient cache. Maybe the right think is to allow configuring a specific `.m2` directory (which could be the default one, if that is what the user wants). This would be more flexible and also feels safer...

**Review by RomainMuller — CHANGES_REQUESTED** [review]

**Comment by RomainMuller** [comment]
Can you perhaps also see to _use_ that configuration flag somewhere (e.g: in one of the flavors of the build test, perhaps)?

File: packages/jsii-pacmak/lib/targets/java.ts [hunk 2]
```diff
@@ -295,8 +295,10 @@ export class JavaBuilder implements TargetBuilder {
             ],
             // Do *not* attempt to ask the user for stuff...
             interactiveMode: false,
-            // Use a non-default local repository to isolate from cached artifacts...
-            localRepository: path.resolve(where, '.m2', 'repository'),
+            // Use a non-default local repository (unless java-custom-cache-path arg is provided) to isolate from cached artifacts...
+            localRepository:
+              this.options.arguments['java-custom-cache-path'] ??
```
[the remark this answers is not in the document]
Comment (agdimech) [reply packages/jsii-pacmak/lib/targets/java.ts]: Fixed.

File: packages/jsii-pacmak/bin/jsii-pacmak.ts [hunk 3]
```diff
@@ -122,6 +122,12 @@ import { VERSION_DESC } from '../lib/version';
       // This is expected to be a path, which should be normalized
       normalize: true,
     })
+    .option('java-custom-cache-path', {
```
[the remark this answers is not in the document]
Comment (agdimech) [reply packages/jsii-pacmak/bin/jsii-pacmak.ts]: Fixed.

**Review by RomainMuller — APPROVED** [review]

**Pull request merged**